src/compiler/checker.ts

```diff
@@ -6011,7 +6018,70 @@ namespace ts {
+                                    //
+                                    //     }
+                                    //     else if (isB(x)) { //  narrow to A
+                                    //
+                                    //     }
+                                    //     else if (isC(x)) { // narrow to A | B
+                                    //
+                                    //     }
+                                    //     else { // type begins with A | B | C
+                                    //
+                                    //          x // is void
+                                    //     }
+                                    //
+                                    if (!(type.flags & TypeFlags.Union)) {
+                                        // Get the original type again because we don't want to store it on a variable higher
+                                        // up to minimize memory allocation.
+                                        let originalType = getTypeOfSymbol(symbol);
+
+                                        if (originalType.flags & TypeFlags.Union &&
+                                            // Use the original type to check the narrowed type instead of the ongoing narrowing type.
+                                            // This is to deal with cases like below:
+                                            //
+                                            //     if (isB(x)) // narrow to A.
+                                            //     else if (isB(x)) // narrow to A
+                                            //
+                                            // Which is the same as:
+                                            //
+                                            //     if (isA(x)) // narrow to A (but also removes A).
+                                            //     else if (isB(x)) // narrow to A
+                                            //
+                                            // But we want only the latter to narrow the type to void.
+                                            //
+                                            type === narrowType(originalType, (<IfStatement>node).expression, /*assumeTrue*/ true) &&
+                                            // If the first narrowed type yield a distinct type(not a union type). And it encounters
+                                            // a narrowed type that yields the same type in some upper branches. It would then narrow
+                                            // the type to void instead of the distinct type. The below check, guards us from narrowing
+                                            // it to void instead of the distinct type.
+                                            //
+                                            //     if (isA(x)) // narrow to void

```

Comment:
I was show casing the error case here. It's an edge case when using `isA(...)` two times. Inside `else if(isA(...))` should expect `x` to be `A` and not `void`.